src/connection.rs

```diff
@@ -163,6 +172,55 @@ impl NetConnection for socket2::Socket {
         Ok(socket)
     }
 
+    fn connect_reusable_nonblocking(
+        local_addr: Self::Addr,
+        remote_addr: Self::Addr,
+    ) -> io::Result<Self> {
+        use std::net::ToSocketAddrs;
+
+        let local_addr = local_addr.to_socket_addrs()?.next().ok_or(io::ErrorKind::InvalidInput)?;
+        let remote_addr =
+            remote_addr.to_socket_addrs()?.next().ok_or(io::ErrorKind::AddrNotAvailable)?;
+        let socket = socket2::Socket::new(
+            socket2::Domain::for_address(local_addr),
+            socket2::Type::STREAM,
+            None,
+        )?;
+        socket.set_nonblocking(true)?;
+        socket.set_reuse_address(true)?;
+        #[cfg(all(unix, not(target_os = "solaris"), not(target_os = "illumos")))]
+        {
+            socket.set_reuse_port(true)?;
+        }
+        socket2::Socket::bind(&socket, &local_addr.into())?;
+
+        match socket2::Socket::connect(&socket, &remote_addr.into()) {
+            Ok(()) => {
+                #[cfg(feature = "log")]
+                log::debug!(target: "netservices", "Connected to {}", remote_addr);
+            }
+            Err(e) if e.raw_os_error() == Some(libc::EINPROGRESS) => {
+                #[cfg(feature = "log")]
+                log::debug!(target: "netservices", "Connecting to {} in a non-blocking way", remote_addr);
+            }
+            Err(e) if e.raw_os_error() == Some(libc::EALREADY) => {
+                #[cfg(feature = "log")]
+                log::error!(target: "netservices", "Can't connect to {}: address already in use", remote_addr);
+                return Err(io::Error::from(io::ErrorKind::AlreadyExists));
+            }
+            Err(e) if e.kind() == io::ErrorKind::WouldBlock => {
+                #[cfg(feature = "log")]
+                log::error!(target: "netservices", "Can't connect to {} in a non-blocking way", remote_addr);
```

Comment:
Why are we ok with skipping this error and returning socket as if it was connected?

Reply:
That logic was copy-pasted from the pre-existing `connect_nonblocking` method. A successful call to `connect` on a non-blocking socket should either return zero or set errno to `EINPROGRESS` (at least according to docs) so I agree it's not correct to ignore `EWOULDBLOCK`, though maybe someone had a reason for writing it like this?

Reply:
Than we can merge that and sort out this question as an independent issue. I used this probably after @cloudhead suggestion (but not sure), so I assume he had some issues with reactor/nakomoto in the past which caused that. Anyway this is the stuff which works well in current radicle, so better not to touch.